srk/src/terminationDTA.ml

```diff
@@ -0,0 +1,599 @@
+   match ExpPolynomial.exponentiate_rational best_dlts_tr_matrix with
+    | None -> failwith "best rational DLTS abstraction produced a matrix with not all rational eigenvalues!"
+    | Some mat -> mat
+  in
+  logf "Printing exp poly:\n%a\n" ExpPolynomial.Matrix.pp exp_poly_mat;
+  exp_poly_mat
+
+type ineq_type = Lt0 | Eq0 | Leq0
+
+(** This data structure is a map with key type (QQ.t, int) and value type QQVector.
+    Ordering on the keys is the lexicographic ordering, and value types
+    form an Abelian group.
+    This is used to store a term in an exponential polynomial. 
+    For example, a term could look like 2^k * (k^3) * (2x + y + z) where
+    k is the symbolic loop counter, and x, y, z are the initial values of
+    variables x, y, z.
+*)
+module BaseDegPairMap = struct
+  type pair = QQ.t * int [@@deriving ord]
+  module QQV = Linear.QQVector
+  module E = SparseMap.Make(struct type t = pair [@@deriving ord] end)(QQV)
+  type t = E.t
+  let empty = E.zero
+
+  (** Put a term into the data structure. 
+      index_pair: the base and the degree of the symbolic loop counter.
+      factor: a multiplicative factor.
+      non_zero_dim: an int that correspond to a particular dimension, and
+          -1 correspond to constants
+      p: the Abelian group map
+  *)
+  let put index_pair factor non_zero_dim p =
+    let qqxvec = QQV.of_list [factor, non_zero_dim] in
+    E.set index_pair (QQV.add (E.get index_pair p) qqxvec) p
+
+  let has_lower_order_than_const base =
+    (QQ.lt base (QQ.of_int 1) )
+
+  let has_const_order base deg =
+    QQ.equal QQ.one base && deg == 0
+
+  (** Rank the items in the data structure according to dominance ordering and
+      produce a formula for sufficient condition for termination.
+      p: the Abelian group map.
+      invariant_terms: these are the linear terms whose closed forms are computed
+          during the symbolic exponentiation of transition matrix. This is 
+          used to interpret the meaning of the vectors as values in the map.
+      lhs_const: the constant that appears in the closed form formula.
+      ineq_type: whether this inequality is <=, = or <.
+  *)
+  let rank srk p invariant_terms lhs_const ineq_type =
+    logf "entering rank";
+    let e = E.enum p in
+    let dim_to_term = fun d -> if d < 0 then mk_one srk else BatList.nth invariant_terms d in
+    let dom_ranked_list = 
+      BatEnum.fold
+        (fun l ((base, deg), qqt) -> 
+           logf "base:%a deg:%d vec:%a\n"
+             QQ.pp base 
+             deg 
+             QQV.pp qqt;
+           (base, deg, qqt) :: l
+        ) 
+        []
+        e
+    in
+    let conditions = 
+      logf "start printing conditions list and formula stems";
+      let conditions_list_final, formula_stem_list, encountered_const_order = 
+        BatList.fold_left
+          (fun (conditions_list, formula_stem, flag) (base, deg, qqt)  ->
+             if has_const_order base deg then
```

Comment:
This is used to deal with one scenario where the LHS has a first-order symbolic constant and the equation to be analyzed is like x0 + 5 = 0. I'm thinking if this can be simplified if I managed to subtract the constant from LHS